staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go

```diff
@@ -111,19 +112,16 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", obj), source, err)
 	}
 
-	var patchType types.PatchType
 	var patch []byte
-	var lookupPatchMeta strategicpatch.LookupPatchMeta
-	var schema oapi.Schema
 	createPatchErrFormat := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
 
-	// Create the versioned struct from the type defined in the restmapping
-	// (which is the API version we'll be submitting the patch to)
-	versionedObject, err := scheme.Scheme.New(p.Mapping.GroupVersionKind)
-	switch {
-	case runtime.IsNotRegisteredError(err):
-		// fall back to generic JSON merge patch
-		patchType = types.MergePatchType
+	patchType, err := p.getPatchType(p.Mapping.GroupVersionKind)
```

Comment:
The general logic here is still super hard to read, and the `getPatchType` only partially helps. I don't know if that's the part that needs to be in a separate function, or if each individual branch should be in a separate function (i.e., building the strategic merge patch from OpenAPI vs from struct vs doing the merge patch.

Reply:
I agree, that's why it takes so much time for me to fix this issue :).

But I'd prefer only changing openapi usage in that PR and I gladly open a cleanup/refactoring for readability changes in followup PRs?

Reply:
Then I'd go the other way around, simplify this code before fixing the bug, so that we have a better understanding of what's going on right now that is wrong, and how it can be fixed. Otherwise it's hard to know that you're doing the right thing.

Reply:
@apelisse I made some refactorings to increase the readability of the code as you suggested. Basically, I split each patch branch into separate functions.